src/prepared_statement.c

```diff
@@ -0,0 +1,544 @@
+	 * need to change any of those).
+	 */
+	diff = strlen(bp.name) - ps->stmt_name_len;
+	buf = pktbuf_temp();
+	if (buf == NULL)
+		goto oom;
+	pktbuf_put_char(buf, pkt->type);
+	pktbuf_put_uint32(buf, pkt->len - diff - 1);	/* length does not include type byte */
+	pktbuf_put_string(buf, bp.portal);
+	pktbuf_put_string(buf, ps->stmt_name);
+	pktbuf_put_bytes(buf, pkt->data.data + pkt->data.read_pos,
+			 pkt->data.write_pos - pkt->data.read_pos);
+	if (!sbuf_queue_packet(&client->sbuf, &server->sbuf, buf))
+		goto oom;
+
+	/*
+	 * Track the Bind command that we send to server and forward the response
+	 * to the client, because they expect one.
+	 */
+	if (!add_outstanding_request(client, 'B', RA_FORWARD))
+		goto oom;
+
+	if (client->packet_cb_state.flag == CB_HANDLE_COMPLETE_PACKET) {
+		/*
+		 * If we used special callback buffering for this packet then there's
+		 * no other bytes that we need to send. This is an exceptional case. It
+		 * only happens when the statement name does not fit in pkt_buf.
+		 */
+		return true;
+	}
+	/*
+	 * For all other cases we tell SBuf to skip over the bytes that were
+	 * in the current packet buffer, but we do still want it to forward all the
+	 * remaining bytes.
+	 */
+	sbuf_prepare_skip_then_send_leftover(&client->sbuf, &server->sbuf, pkt->data.write_pos, pkt->len);
+	return true;
+
+oom:
+	disconnect_client(client, true, "out of memory");
+	/*
+	 * We also disconnect the server, because we probably messed with its state
+	 * at this prepared statement state at this point. And rolling that back is
+	 * hard.
+	 */
+	disconnect_server(client->link, true, "out of memory");
+	return false;
+}
+
+/*
+ * Handle a Describe packet for a named prepared statement
+ *
+ * Rewrites the given Describe packet to use the server-side statement name
+ * instead of the client-side one. If the statement is not yet prepared on the
+ * server, then we first send a Parse command to the server for the prepared
+ * statement in question.
+ */
+bool handle_describe_command(PgSocket *client, PktHdr *pkt)
+{
+	PgSocket *server = client->link;
+	PgDescribePacket dp;
+	PgClientPreparedStatement *client_ps = NULL;
+	PgPreparedStatement *ps;
+	bool res;
+
+	Assert(server);
+
+	if (!unmarshall_describe_packet(client, pkt, &dp) || dp.type != 'S')
+		return false;
+
+	client_ps = get_client_prepared_statement(client, dp.name);
+	if (!client_ps)
+		return false;
+	ps = client_ps->ps;
+
+	if (!ensure_statement_is_prepared_on_server(server, ps))
+		goto oom;
+
+	slog_debug(client, "handle_describe_command: mapped statement '%s' (query '%s') to '%s'",
+		   dp.name, ps->query_and_parameters, ps->stmt_name);
+
+	/*
+	 * Track the Describe command that we send to server and forward the
+	 * response to the client, because they expect one.
+	 */
+	if (!add_outstanding_request(client, 'D', RA_FORWARD))
+		goto oom;
+
+	QUEUE_DescribeStmt(res, client, server, ps->stmt_name);
```

Comment:
```suggestion
	sbuf_prepare_skip(&client->sbuf, pkt->len);
	QUEUE_DescribeStmt(res, client, server, ps->stmt_name);
```

Reply:
Should we add `sbuf_prepare_skip` to handle_describe_command?
With the proposed patch the test above is passed.
What I see i debugger and in strace is that pgbouncer `receives` multiple describe packets (although sqlalchemy didn't send them) and so sends multiples replies which ar not expected by sqlalchemy.

Reply:
Looks good. I think we should do the same as we do for `handle_parse_command`:
```c
	/*
	 * Now we need to make sure the original packet is not sent to the server.
	 */
	if (client->packet_cb_state.flag == CB_HANDLE_COMPLETE_PACKET) {
		/*
		 * If we used special callback packet buffering, then we don't need to
		 * do anything. Because the callback already "consumed" the data from
		 * the SBuf.
		 */
		return true;
	}
	/*
	 * If we used the regular packet buffering, then we do still need to tell
	 * SBuf that we handled the packet.
	 */
	sbuf_prepare_skip(&client->sbuf, pkt->len);
```

We should probably do the same for the Close message. Right now that is also sent twice afaict. Probably best to make the snippet above a function that we call in those 3 places.

Reply:
`handle_prepare_close` already does `stuff_prepare_skip`